Readme.markdown
# PostgreSQL Migrations for Haskell

[![Build Status](https://api.travis-ci.org/ameingast/postgresql-simple-migration.png)](https://travis-ci.org/ameingast/postgresql-simple-migration)

Welcome to postgresql-simple-migrations, a tool for helping you with
PostgreSQL schema migrations.

This project is an open-source database migration tool. It favors simplicity
over configuration.

It is implemented in Haskell and uses the (excellent) postgresql-simple
library to communicate with PostgreSQL.

It comes in two flavors: a library that features an easy to use Haskell
API and as a standalone application.

Database migrations can be written in SQL (in this case PostgreSQL-sql)
or in Haskell.

## Why?
Database migrations should not be hard. They should be under version control
and documented both in your production systems and in your project files.

## What?
This library executes SQL/Haskell migration scripts and keeps track of their
meta information.

Scripts are be executed exactly once and any changes to scripts will cause
a run-time error notifying you of a corrupted database.

The meta information consists of:
* an MD5 checksum of the executed script to make sure already existing
  scripts cannot be modified in your production system.
* a time-stamp of the date of execution so you can easily track when a change
  happened.

This library also supports migration validation so you can ensure (some)
correctness before your application logic kicks in.

## How?
This utility can be used in two ways: embedded in your Haskell program or as
a standalone binary.

### Standalone
The standalone program supports file-based migrations. To execute all SQL-files
in a directory $BASE\_DIR, execute the following command to initialize the database
in a first step.

```bash
CON="host=$host dbname=$db user=$user password=$pw"
./dist/build/migrate/migrate init $CON
./dist/build/migrate/migrate migrate $CON $BASE_DIR
```

To validate already executed scripts, execute the following:
```bash
CON="host=$host dbname=$db user=$user password=$pw"
./dist/build/migrate/migrate init $CON
./dist/build/migrate/migrate validate $CON $BASE_DIR
```

For more information about the PostgreSQL connection string, see:
[libpq-connect](http://www.postgresql.org/docs/9.3/static/libpq-connect.html).

### Library
The library supports more actions than the standalone program.

Initializing the database:

```haskell
main :: IO ()
main = do
    let url = "host=$host dbname=$db user=$user password=$pw"
    con <- connectPostgreSQL (BS8.pack url)
    withTransaction con $ runMigration $
        MigrationContext MigrationInitialization True con
```

For file-based migrations, the following snippet can be used:

```haskell
main :: IO ()
main = do
    let url = "host=$host dbname=$db user=$user password=$pw"
    let dir = "."
    con <- connectPostgreSQL (BS8.pack url)
    withTransaction con $ runMigration $
        MigrationContext (MigrationDirectory dir) True con
```

To run Haskell-based migrations, use this:

```haskell
main :: IO ()
main = do
    let url = "host=$host dbname=$db user=$user password=$pw"
    let name = "my script"
    let script = "create table users (email varchar not null)";
    con <- connectPostgreSQL (BS8.pack url)
    withTransaction con $ runMigration $
        MigrationContext (MigrationScript name script) True con
```

Validations wrap _MigrationCommands_. This means that you can re-use all
MigrationCommands to perform a read-only validation of your migrations.

To perform a validation on a directory-based migration, you can use the
following code:

```haskell
main :: IO ()
main = do
    let url = "host=$host dbname=$db user=$user password=$pw"
    con <- connectPostgreSQL (BS8.pack url)
    withTransaction con $ runMigration $ MigrationContext
        (MigrationValidation (MigrationDirectory dir)) True con
```

Database migrations should always be performed in a transactional context.

The standalone binary takes care of proper transaction handling automatically.

The library does not make any assumptions about the current transactional state
of the system. This means that the caller of the library has to take care of
opening/closing/rolling-back transactions. This way you can execute multiple
migration-commands or validations in sequence while still staying in the
transaction you opened.

The tests make use of this. After executing all migration-tests, the
transaction is rolled back.

## Compilation and Tests
The program is built with the _cabal_ build system. The following command
builds the library, the standalone binary and the test package.

```bash
cabal configure --enable-tests && cabal build -j
```

To execute the tests, you need a running PostgreSQL server with an empty
database called _test_. Tests are executed through cabal as follows:

```bash
cabal configure --enable-tests && cabal test
```

To build the project in a cabal sandbox, use the following code:

```bash
cabal sandbox init
cabal install -j --only-dependencies --enable-tests --disable-documentation
cabal configure --enable-tests
cabal test
```

To remove the generated cabal sandbox, use:
```bash
cabal sandbox delete
```

## To Do
* Collect executed scripts and check if already executed scripts have been
  deleted.


Changelog.markdown
# Changelog

## 0.1.15.0
* Bumped dependencies

## 0.1.14.0
* Bumped dependencies

## 0.1.13.1
* Bumped dependencies

## 0.1.13.0
* Bumped dependencies

## 0.1.12.0
* Support for GHC 8.4

## 0.1.11.0
* Improved documentation
* Fixed exists_table

## 0.1.10.1
* Fixed hackage warnings

## 0.1.10.0
* Relaxed time bounds

## 0.1.9.0
* Bumped dependencies

## 0.1.8.0
* Added MigrationCommands allowing sequencing of migrations in the Haskell API
* Derived more datatypes for MigrationResult
* Bumped dependencies

## 0.1.7.0
* Propagate migration and validation result to application exit code

## 0.1.6.0
* Support for GHC 8

## 0.1.5.0
* Bumped dependencies

## 0.1.4.0
* Improved error logging in standalone binary

## 0.1.3.0
* Better transaction handling
* Improved documentation

## 0.1.2.0
* Moved Util module
* Improved documentation

## 0.1.1.0
* Support for schema validations.
* Improved Haskell API

## 0.1.0.0
* Support for file-based and Haskell migrations.


test/Main.hs
-- |
-- Module      : Main
-- Copyright   : (c) 2014 Andreas Meingast <ameingast@gmail.com>
--
-- License     : BSD-style
-- Maintainer  : ameingast@gmail.com
-- Stability   : experimental
-- Portability : GHC
--
-- The test entry-point for postgresql-simple-migration.

{-# LANGUAGE CPP               #-}
{-# LANGUAGE LambdaCase        #-}
{-# LANGUAGE OverloadedStrings #-}

module Main
    ( main
    ) where

import           Database.PostgreSQL.Simple               (connectPostgreSQL)
import           Database.PostgreSQL.Simple.MigrationTest (metaMigrationSpec,
                                                           migrationSpec)
import           Database.PostgreSQL.Simple.Util          (withTransactionRolledBack)
import           Test.Hspec                               (hspec)

main :: IO ()
main = do
    con <- connectPostgreSQL "dbname=test"
    withTransactionRolledBack con (hspec (metaMigrationSpec con))
    withTransactionRolledBack con (hspec (migrationSpec con))


test/Database/PostgreSQL/Simple/ParserTest.hs
-- |
-- Module      : Database.PostgreSQL.Simple.MigrationTest
-- Copyright   : (c) 2014 Andreas Meingast <ameingast@gmail.com>
--
-- License     : BSD-style
-- Maintainer  : ameingast@gmail.com
-- Stability   : experimental
-- Portability : GHC
--
-- A collection of postgresql-simple-migration specifications.

{-# LANGUAGE QuasiQuotes #-}

module Database.PostgreSQL.Simple.ParserTest where

import           Control.Applicative               ((<$>))
import           Data.ByteString                   (ByteString)
import           Data.Monoid                       (mconcat)
import           Data.String.QQ
import           Database.PostgreSQL.Simple.Parser (toQueries)
import           Database.PostgreSQL.Simple.Types  (Query (..))
import           Test.Hspec                        (Spec, describe, it,
                                                    shouldBe)

query1, query2, query3, complexQ1, complexQ2, finalQuery :: ByteString
query1 = [s|CREATE TABLE test (test_item TEXT);|]
query2 = [s|INSERT INTO test values ('testing');|]
query3 = [s|DROP TABLE test;|]
complexQ1 = [s|INSERT INTO TABLE test values
                 ('Some;Semicolons;inside;quotes;');|]
complexQ2 = [s|INSERT INTO TABLE test values
                 ('Semicolons;\'inside\';a;complex;quote');|]
finalQuery = [s|SELECT 2 + 2 AS basic|]


parserSpec :: Spec
parserSpec = describe "Parser" $ do
  it "Will separate queries based on semicolon" $
    let q = toQueries (mconcat [query1, query2, query3])
    in length q `shouldBe` 3

  it "Will parse complete queries in the right order" $
    let q = toQueries (mconcat [query1, query2, query3])
    in q `shouldBe` (Query <$> [query1, query2, query3])

  it "Will not separate queries on semicolons inside quotes" $
    let q = toQueries (mconcat [query1, complexQ1, query2])
    in length q `shouldBe` 3

  it "Will understand denormalized quotes" $
    let q = toQueries (mconcat [query1, complexQ2, query2])
    in length q `shouldBe` 3
  it "Will be able to process a final query without a semicolon" $
    let q = toQueries (mconcat [query1, query2, finalQuery])
    in q `shouldBe` (Query <$> [query1, query2, finalQuery])


test/Database/PostgreSQL/Simple/MigrationTest.hs
-- |
-- Module      : Database.PostgreSQL.Simple.MigrationTest
-- Copyright   : (c) 2014 Andreas Meingast <ameingast@gmail.com>
--
-- License     : BSD-style
-- Maintainer  : ameingast@gmail.com
-- Stability   : experimental
-- Portability : GHC
--
-- A collection of postgresql-simple-migration specifications.

{-# LANGUAGE CPP               #-}
{-# LANGUAGE LambdaCase        #-}
{-# LANGUAGE OverloadedStrings #-}

module Database.PostgreSQL.Simple.MigrationTest where

import           Database.PostgreSQL.Simple           (Connection, execute_)
import           Database.PostgreSQL.Simple.Migration (MigrationCommand (..),
                                                       MigrationContext (..),
                                                       MigrationResult (..),
                                                       SchemaMigration (..),
                                                       getMigrations,
                                                       runMigration)
import           Database.PostgreSQL.Simple.Util      (existsTable, getDataType)
import           Test.Hspec                           (Spec, describe, it,
                                                       shouldBe)
metaMigrationSpec:: Connection -> Spec
metaMigrationSpec con = describe "Meta Migrations" $ do
    it "asserts that the schema_migrations table does not exist" $ do
        r <- existsTable con "schema_migrations"
        r `shouldBe` False

    it "creates an old version of the schema_migrations table for testing" $ do
        -- This is an old version of initializeSchema, notice the type of
        -- executed_at.
        _ <- execute_ con $ mconcat
            [ "create table if not exists schema_migrations "
            , "( filename varchar(512) not null"
            , ", checksum varchar(32) not null"
            , ", executed_at timestamp without time zone not null default now() "
            , ");"
            ]

        r <- existsTable con "schema_migrations"
        r `shouldBe` True

    it "runs a validation, which automatically meta-migrates" $ do
        r <- runMigration $ MigrationContext
            (MigrationValidation MigrationInitialization) False con
        r `shouldBe` MigrationSuccess
        t <- getDataType con "schema_migrations" "executed_at"
        t `shouldBe` Just "timestamp with time zone"

migrationSpec:: Connection -> Spec
migrationSpec con = describe "Migrations" $ do
    let migrationScript = MigrationScript "test.sql" q
    let migrationScriptAltered = MigrationScript "test.sql" ""
    let migrationDir = MigrationDirectory "share/test/scripts"
    let migrationFile = MigrationFile "s.sql" "share/test/script.sql"

    it "asserts that the schema_migrations table does not exist" $ do
        r <- existsTable con "schema_migrations"
        r `shouldBe` False

    it "validates an initialization on an empty database" $ do
        r <- runMigration $ MigrationContext
            (MigrationValidation MigrationInitialization) False con
        r `shouldBe` MigrationError "No such table: schema_migrations"

    it "initializes a database" $ do
        r <- runMigration $ MigrationContext MigrationInitialization False con
        r `shouldBe` MigrationSuccess

    it "creates the schema_migrations table" $ do
        r <- existsTable con "schema_migrations"
        r `shouldBe` True

    it "executes a migration script" $ do
        r <- runMigration $ MigrationContext migrationScript False con
        r `shouldBe` MigrationSuccess

    it "creates the table from the executed script" $ do
        r <- existsTable con "t1"
        r `shouldBe` True

    it "skips execution of the same migration script" $ do
        r <- runMigration $
            MigrationContext migrationScript False con
        r `shouldBe` MigrationSuccess

    it "reports an error on a different checksum for the same script" $ do
        r <- runMigration $ MigrationContext migrationScriptAltered False con
        r `shouldBe` MigrationError "test.sql"

    it "executes migration scripts inside a folder" $ do
        r <- runMigration $ MigrationContext migrationDir False con
        r `shouldBe` MigrationSuccess

    it "creates the table from the executed scripts" $ do
        r <- existsTable con "t2"
        r `shouldBe` True

    it "executes a file based migration script" $ do
        r <- runMigration $ MigrationContext migrationFile False con
        r `shouldBe` MigrationSuccess

    it "creates the table from the executed scripts" $ do
        r <- existsTable con "t3"
        r `shouldBe` True

    it "validates initialization" $ do
        r <- runMigration $ MigrationContext
            (MigrationValidation MigrationInitialization) False con
        r `shouldBe` MigrationSuccess

    it "validates an executed migration script" $ do
        r <- runMigration $ MigrationContext
            (MigrationValidation migrationScript) False con
        r `shouldBe` MigrationSuccess

    it "validates all scripts inside a folder" $ do
        r <- runMigration $ MigrationContext
            (MigrationValidation migrationDir) False con
        r `shouldBe` MigrationSuccess

    it "validates an executed migration file" $ do
        r <- runMigration $ MigrationContext
            (MigrationValidation migrationFile) False con
        r `shouldBe` MigrationSuccess

    it "gets a list of executed migrations" $ do
        r <- getMigrations con
        map schemaMigrationName r `shouldBe` ["test.sql", "1.sql", "s.sql"]

    where
        q = "create table t1 (c1 varchar);"



src/Database/PostgreSQL/Simple/Util.hs
-- |
-- Module      : Database.PostgreSQL.Simple.Util
-- Copyright   : (c) 2014 Andreas Meingast <ameingast@gmail.com>
--
-- License     : BSD-style
-- Maintainer  : ameingast@gmail.com
-- Stability   : experimental
-- Portability : GHC
--
-- A collection of utilites for database migrations.

{-# LANGUAGE CPP               #-}
{-# LANGUAGE LambdaCase        #-}
{-# LANGUAGE OverloadedStrings #-}

module Database.PostgreSQL.Simple.Util
    ( existsTable
    , withTransactionRolledBack
    ) where

import           Control.Exception          (finally)
import           Database.PostgreSQL.Simple (Connection, Only (..), begin,
                                             query, rollback)
import           GHC.Int                    (Int64)

-- | Checks if the table with the given name exists in the database.
existsTable :: Connection -> String -> IO Bool
existsTable con table =
    fmap checkRowCount (query con q (Only table) :: IO [[Int64]])
    where
        q = "select count(relname) from pg_class where relname = ?"

        checkRowCount :: [[Int64]] -> Bool
        checkRowCount ((1:_):_) = True
        checkRowCount _         = False

-- | Executes the given IO monad inside a transaction and performs a roll-back
-- afterwards (even if exceptions occur).
withTransactionRolledBack :: Connection -> IO a -> IO a
withTransactionRolledBack con f =
    begin con >> finally f (rollback con)


src/Database/PostgreSQL/Simple/Parser.hs
-- |
-- Module      : Database.PostgreSQL.Simple.Util
-- Copyright   : (c) 2014 Andreas Meingast <ameingast@gmail.com>
--
-- License     : BSD-style
-- Maintainer  : ameingast@gmail.com
-- Stability   : experimental
-- Portability : GHC
--
-- A collection of utilites for database migrations.

{-# LANGUAGE RecordWildCards #-}

module Database.PostgreSQL.Simple.Parser
  (
    toQueries
  ) where

import qualified Data.ByteString                  as BS (ByteString, foldl)
import qualified Data.ByteString.Builder          as BSB
import qualified Data.ByteString.Lazy             as BSI (ByteString, toStrict)
import           Data.Monoid                      (mempty, (<>))
import           Data.Word
import           Database.PostgreSQL.Simple.Types (Query (..))

backslash, singleQuote, semicolon, lineFeed, carriageReturn :: Word8
backslash = 0x5c
singleQuote = 0x27
semicolon = 0x3b
lineFeed = 0x0a
carriageReturn = 0x0d

data ParsingContext = InQuote
                    | Escaping
                    | OutQuote

data ParsedQueries =
  Parsed { current  :: BSB.Builder
         , previous :: [BSI.ByteString] }

addToCurrent :: Word8 -> ParsedQueries -> ParsedQueries
addToCurrent c p@Parsed{..} = p { current = current <> BSB.word8 c }

finishCurrent :: ParsedQueries -> ParsedQueries
finishCurrent p@Parsed{..} =
  let builded = BSB.toLazyByteString current
  in if builded == mempty
     then p
     else p { current = mempty
            , previous = BSB.toLazyByteString current:previous }

type ParsingState = (ParsingContext, ParsedQueries)

-- | Cut a migration file into a series of queries so they can be isolated and
-- handled sequentially, helping identify which one failed if any.
toQueries :: BS.ByteString -> [Query]
toQueries =
  let parsed = BS.foldl parseChar (OutQuote, Parsed mempty [])
      queries = reverse . previous . finishCurrent . snd . parsed
  in fmap (Query . BSI.toStrict) . queries

parseChar :: ParsingState -> Word8 -> ParsingState
parseChar (InQuote, p) c
  | c == backslash = (Escaping, addToCurrent c p)
  | c == singleQuote = (OutQuote, addToCurrent c p)
  | otherwise = (InQuote, addToCurrent c p)
parseChar (Escaping, p) c = (InQuote, addToCurrent c p)
parseChar (OutQuote, p) c
  | c == singleQuote = (InQuote, addToCurrent c p)
  | c == semicolon = (OutQuote, finishCurrent . addToCurrent semicolon $ p)
  | c == semicolon = (OutQuote, finishCurrent . addToCurrent semicolon $ p)
  | c `elem` [lineFeed, carriageReturn] = (OutQuote, p)
  | otherwise = (OutQuote, addToCurrent c p)


src/Database/PostgreSQL/Simple/Migration.hs
-- |
-- Module      : Database.PostgreSQL.Simple.Migration
-- Copyright   : (c) 2014 Andreas Meingast <ameingast@gmail.com>
--
-- License     : BSD-style
-- Maintainer  : ameingast@gmail.com
-- Stability   : experimental
-- Portability : GHC
--
-- A migration library for postgresql-simple.
--
-- For usage, see Readme.markdown.

{-# LANGUAGE CPP               #-}
{-# LANGUAGE DeriveFoldable    #-}
{-# LANGUAGE DeriveFunctor     #-}
{-# LANGUAGE DeriveTraversable #-}
{-# LANGUAGE LambdaCase        #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE NamedFieldPuns    #-}

module Database.PostgreSQL.Simple.Migration
    (
    -- * Migration actions
      runMigration'
    , runMigration
    , runMigrations
    , runMigrations'
    , sequenceMigrations

    -- * Migration types
    , MigrationContext(..)
    , MigrationContext'(..)
    , MigrationCommand(..)
    , MigrationResult(..)
    , ScriptName
    , Checksum

    -- * Migration result actions
    , getMigrations
    , getMigrations'

    -- * Migration result types
    , SchemaMigration(..)
    ) where

#if __GLASGOW_HASKELL__ < 710
import           Control.Applicative                ((<$>), (<*>))
#endif
import           Control.Monad                      (void, when)
import qualified Crypto.Hash.MD5                    as MD5 (hash)
import qualified Data.ByteString                    as BS (ByteString, readFile)
import qualified Data.ByteString.Char8              as BS8 (unpack)
import qualified Data.ByteString.Base64             as B64 (encode)
import           Data.Foldable                      (Foldable)
import           Data.List                          (isPrefixOf, sort)
import           Data.Traversable                   (Traversable)
#if __GLASGOW_HASKELL__ < 710
import           Data.Monoid                        (Monoid (..))
#endif
import           Data.Time                          (LocalTime)
import           Database.PostgreSQL.Simple         (Connection, Only (..),
                                                     execute, execute_, query,
                                                     query_)
import           Database.PostgreSQL.Simple.FromRow (FromRow (..), field)
import           Database.PostgreSQL.Simple.ToField (ToField (..))
import           Database.PostgreSQL.Simple.ToRow   (ToRow (..))
import           Database.PostgreSQL.Simple.Types   (Query (..))
import           Database.PostgreSQL.Simple.Util    (existsTable)
import           System.Directory                   (getDirectoryContents)

-- | Executes migrations inside the provided 'MigrationContext'.
--
-- Returns 'MigrationSuccess' if the provided 'MigrationCommand' executes
-- without error. If an error occurs, execution is stopped and
-- a 'MigrationError' is returned.
--
-- It is recommended to wrap 'runMigration' inside a database transaction.
runMigration :: MigrationContext -> IO (MigrationResult String)
runMigration (MigrationContext cmd verbose con) = 
  runMigration' (MigrationContext' cmd verbose con "schema_migrations")

runMigration' :: MigrationContext' -> IO (MigrationResult String)
runMigration' (MigrationContext' cmd verbose con tableName) = case cmd of
    MigrationInitialization ->
        initializeSchema con tableName verbose >> return MigrationSuccess
    MigrationDirectory path ->
        executeDirectoryMigration con tableName verbose path
    MigrationScript name contents ->
        executeMigration con tableName verbose name contents
    MigrationFile name path ->
        executeMigration con tableName verbose name =<< BS.readFile path
    MigrationValidation validationCmd ->
        executeValidation con tableName verbose validationCmd
    MigrationCommands commands ->
        runMigrations' verbose con commands tableName

-- | Execute a sequence of migrations
--
-- Returns 'MigrationSuccess' if all of the provided 'MigrationCommand's
-- execute without error. If an error occurs, execution is stopped and the
-- 'MigrationError' is returned.
--
-- It is recommended to wrap 'runMigrations' inside a database transaction.
runMigrations
    :: Bool
       -- ^ Run in verbose mode
    -> Connection
       -- ^ The postgres connection to use
    -> [MigrationCommand]
       -- ^ The commands to run
    -> IO (MigrationResult String)
runMigrations verbose con commands = runMigrations' verbose con commands "schema_migrations"

-- | Execute a sequence of migrations
--
-- Returns 'MigrationSuccess' if all of the provided 'MigrationCommand's
-- execute without error. If an error occurs, execution is stopped and the
-- 'MigrationError' is returned.
--
-- It is recommended to wrap 'runMigrations' inside a database transaction.
runMigrations'
    :: Bool
       -- ^ Run in verbose mode
    -> Connection
       -- ^ The postgres connection to use
    -> [MigrationCommand]
       -- ^ The commands to run
    -> BS.ByteString
       -- ^ The schema_migrations table name
    -> IO (MigrationResult String)
runMigrations' verbose con commands tableName =
    sequenceMigrations [runMigration' (MigrationContext' c verbose con tableName) | c <- commands]

-- | Run a sequence of contexts, stopping on the first failure
sequenceMigrations :: Monad m => [m (MigrationResult e)] -> m (MigrationResult e)
sequenceMigrations = \case
    []   -> return MigrationSuccess
    c:cs -> do
        r <- c
        case r of
          MigrationError s -> return (MigrationError s)
          MigrationSuccess -> sequenceMigrations cs

-- | Executes all SQL-file based migrations located in the provided 'dir'
-- in alphabetical order.
executeDirectoryMigration :: Connection -> BS.ByteString -> Bool -> FilePath -> IO (MigrationResult String)
executeDirectoryMigration con tableName verbose dir =
    scriptsInDirectory dir >>= go
    where
        go fs = sequenceMigrations (executeMigrationFile <$> fs)
        executeMigrationFile f = executeMigration con tableName verbose f =<< BS.readFile (dir ++ "/" ++ f)

-- | Lists all files in the given 'FilePath' 'dir' in alphabetical order.
scriptsInDirectory :: FilePath -> IO [String]
scriptsInDirectory dir =
    fmap (sort . filter (\x -> not $ "." `isPrefixOf` x))
        (getDirectoryContents dir)

-- | Executes a generic SQL migration for the provided script 'name' with
-- content 'contents'.
executeMigration :: Connection -> BS.ByteString -> Bool -> ScriptName -> BS.ByteString -> IO (MigrationResult String)
executeMigration con tableName verbose name contents = do
    let checksum = md5Hash contents
    checkScript con tableName name checksum >>= \case
        ScriptOk -> do
            when verbose $ putStrLn $ "Ok:\t" ++ name
            return MigrationSuccess
        ScriptNotExecuted -> do
            void $ execute_ con (Query contents)
            void $ execute con q (name, checksum)
            when verbose $ putStrLn $ "Execute:\t" ++ name
            return MigrationSuccess
        ScriptModified { actual, expected } -> do
            when verbose $ putStrLn
              $ "Fail:\t" ++ name
              ++ "\n" ++ scriptModifiedErrorMessage expected actual
            return (MigrationError name)
    where
        q = "insert into " <> Query tableName <> "(filename, checksum) values(?, ?)"

-- | Initializes the database schema with a helper table containing
-- meta-information about executed migrations.
initializeSchema :: Connection -> BS.ByteString -> Bool -> IO ()
initializeSchema con tableName verbose = do
    when verbose $ putStrLn "Initializing schema"
    void $ execute_ con $ mconcat
        [ "create table if not exists " <> Query tableName <> " "
        , "( filename varchar(512) not null"
        , ", checksum varchar(32) not null"
        , ", executed_at timestamp without time zone not null default now() "
        , ");"
        ]

-- | Validates a 'MigrationCommand'. Validation is defined as follows for these
-- types:
--
-- * 'MigrationInitialization': validate the presence of the meta-information
-- table.
-- * 'MigrationDirectory': validate the presence and checksum of all scripts
-- found in the given directory.
-- * 'MigrationScript': validate the presence and checksum of the given script.
-- * 'MigrationFile': validate the presence and checksum of the given file.
-- * 'MigrationValidation': always succeeds.
-- * 'MigrationCommands': validates all the sub-commands stopping at the first failure.
executeValidation :: Connection -> BS.ByteString -> Bool -> MigrationCommand -> IO (MigrationResult String)
executeValidation con tableName' verbose cmd = 
  let tableName = BS8.unpack tableName' in
  case cmd of
    MigrationInitialization ->
        existsTable con tableName >>= \r -> return $ if r
            then MigrationSuccess
            else MigrationError $ "No such table: " <> tableName
    MigrationDirectory path ->
        scriptsInDirectory path >>= goScripts path
    MigrationScript name contents ->
        validate name contents
    MigrationFile name path ->
        validate name =<< BS.readFile path
    MigrationValidation _ ->
        return MigrationSuccess
    MigrationCommands cs ->
        sequenceMigrations (executeValidation con tableName' verbose <$> cs)
    where
        validate name contents =
            checkScript con tableName' name (md5Hash contents) >>= \case
                ScriptOk -> do
                    when verbose $ putStrLn $ "Ok:\t" ++ name
                    return MigrationSuccess
                ScriptNotExecuted -> do
                    when verbose $ putStrLn $ "Missing:\t" ++ name
                    return (MigrationError $ "Missing: " ++ name)
                ScriptModified { expected, actual } -> do
                    when verbose $ putStrLn
                      $ "Checksum mismatch:\t" ++ name
                      ++ "\n" ++ scriptModifiedErrorMessage expected actual
                    return (MigrationError $ "Checksum mismatch: " ++ name)

        goScripts path xs = sequenceMigrations (goScript path <$> xs)
        goScript path x = validate x =<< BS.readFile (path ++ "/" ++ x)

-- | Checks the status of the script with the given name 'name'.
-- If the script has already been executed, the checksum of the script
-- is compared against the one that was executed.
-- If there is no matching script entry in the database, the script
-- will be executed and its meta-information will be recorded.
checkScript :: Connection -> BS.ByteString -> ScriptName -> Checksum -> IO CheckScriptResult
checkScript con tableName name fileChecksum =
    query con q (Only name) >>= \case
        [] ->
            return ScriptNotExecuted
        Only dbChecksum:_ | fileChecksum == dbChecksum ->
            return ScriptOk
        Only dbChecksum:_ ->
            return (ScriptModified {
                       expected = dbChecksum,
                       actual = fileChecksum
                    })
    where
        q = mconcat
            [ "select checksum from " <> Query tableName <> " "
            , "where filename = ? limit 1"
            ]

-- | Calculates the MD5 checksum of the provided bytestring in base64
-- encoding.
md5Hash :: BS.ByteString -> Checksum
md5Hash = B64.encode . MD5.hash

-- | The checksum type of a migration script.
type Checksum = BS.ByteString

-- | The name of a script. Typically the filename or a custom name
-- when using Haskell migrations.
type ScriptName = String

-- | 'MigrationCommand' determines the action of the 'runMigration' script.
data MigrationCommand
    = MigrationInitialization
    -- ^ Initializes the database with a helper table containing meta
    -- information.
    | MigrationDirectory FilePath
    -- ^ Executes migrations based on SQL scripts in the provided 'FilePath'
    -- in alphabetical order.
    | MigrationFile ScriptName FilePath
    -- ^ Executes a migration based on script located at the provided
    -- 'FilePath'.
    | MigrationScript ScriptName BS.ByteString
    -- ^ Executes a migration based on the provided bytestring.
    | MigrationValidation MigrationCommand
    -- ^ Validates that the provided MigrationCommand has been executed.
    | MigrationCommands [MigrationCommand]
    -- ^ Performs a series of 'MigrationCommand's in sequence.
    deriving (Show, Eq, Read, Ord)

#if __GLASGOW_HASKELL__ >= 804
instance Semigroup MigrationCommand where
    (<>) = mappend
#endif

instance Monoid MigrationCommand where
    mempty = MigrationCommands []
    mappend (MigrationCommands xs) (MigrationCommands ys) = MigrationCommands (xs ++ ys)
    mappend (MigrationCommands xs) y = MigrationCommands (xs ++ [y])
    mappend x (MigrationCommands ys) = MigrationCommands (x : ys)
    mappend x y = MigrationCommands [x, y]

-- | A sum-type denoting the result of a single migration.
data CheckScriptResult
    = ScriptOk
    -- ^ The script has already been executed and the checksums match.
    -- This is good.
    | ScriptModified { expected :: Checksum, actual :: Checksum }
    -- ^ The script has already been executed and there is a checksum
    -- mismatch. This is bad.
    | ScriptNotExecuted
    -- ^ The script has not been executed, yet. This is good.
    deriving (Show, Eq, Read, Ord)

scriptModifiedErrorMessage :: Checksum -> Checksum -> [Char]
scriptModifiedErrorMessage expected actual =
  "expected: " ++ show expected ++ "\nhash was: " ++ show actual

-- | A sum-type denoting the result of a migration.
data MigrationResult a
    = MigrationError a
    -- ^ There was an error in script migration.
    | MigrationSuccess
    -- ^ All scripts have been executed successfully.
    deriving (Show, Eq, Read, Ord, Functor, Foldable, Traversable)

-- | The 'MigrationContext' provides an execution context for migrations.
data MigrationContext = MigrationContext
    { migrationContextCommand    :: !MigrationCommand
    -- ^ The action that will be performed by 'runMigration'
    , migrationContextVerbose    :: !Bool
    -- ^ Verbosity of the library.
    , migrationContextConnection :: !Connection
    -- ^ The PostgreSQL connection to use for migrations.
    }

-- | The 'MigrationContext'' provides an execution context for migrations, with additional options to MigrationContext
data MigrationContext' = MigrationContext'
    { migrationContextCommand'    :: !MigrationCommand
    -- ^ The action that will be performed by 'runMigration'
    , migrationContextVerbose'    :: !Bool
    -- ^ Verbosity of the library.
    , migrationContextConnection' :: !Connection
    -- ^ The PostgreSQL connection to use for migrations.
    , migrationTableName :: !BS.ByteString
    -- ^ The name of the table that stores the migrations
    }
--
-- | Produces a list of all executed 'SchemaMigration's.
getMigrations :: Connection -> IO [SchemaMigration]
getMigrations con = getMigrations' con "schema_migrations"

-- | Produces a list of all executed 'SchemaMigration's.
getMigrations' :: Connection -> BS.ByteString -> IO [SchemaMigration]
getMigrations' con tableName = query_ con q
    where q = mconcat
            [ "select filename, checksum, executed_at "
            , "from " <> Query tableName <> " order by executed_at asc"
            ]

-- | A product type representing a single, executed 'SchemaMigration'.
data SchemaMigration = SchemaMigration
    { schemaMigrationName       :: BS.ByteString
    -- ^ The name of the executed migration.
    , schemaMigrationChecksum   :: Checksum
    -- ^ The calculated MD5 checksum of the executed script.
    , schemaMigrationExecutedAt :: LocalTime
    -- ^ A timestamp without timezone of the date of execution of the script.
    } deriving (Show, Eq, Read)

instance Ord SchemaMigration where
    compare (SchemaMigration nameLeft _ _) (SchemaMigration nameRight _ _) =
        compare nameLeft nameRight

instance FromRow SchemaMigration where
    fromRow = SchemaMigration <$>
        field <*> field <*> field

instance ToRow SchemaMigration where
    toRow (SchemaMigration name checksum executedAt) =
       [toField name, toField checksum, toField executedAt]
